sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py

```diff
@@ -94,6 +101,84 @@ def on_request(self, request):
         self._update_headers(request.http_request.headers, self._token.token)
 
 
+class ChallengeAuthenticationPolicy(HTTPPolicy):
+    """Base class for policies that authorize requests with bearer tokens and expect authentication challenges
+
+    :param ~azure.core.credentials.TokenCredential credential: an object which can provide access tokens, such as a
+        credential from :mod:`azure.identity`
+    :param str scopes: required authentication scopes
+    """
+
+    def __init__(self, credential, *scopes, **kwargs):  # pylint:disable=unused-argument
+        # type: (TokenCredential, *str, **Any) -> None
+        super(ChallengeAuthenticationPolicy, self).__init__()
+        self._scopes = scopes
+        self._credential = credential
+        self._token = None  # type: Optional[AccessToken]
+
+    def _need_new_token(self):
+        # type: () -> bool
+        return not self._token or self._token.expires_on - time.time() < 300
+
+    def authorize_request(self, request, *scopes, **kwargs):
```

Comment:
What is the typical way for users to use the policy? Not in the pipeline?

Why in addition to send, we also need to expose authorize_request & on_request?

Reply:
In a pipeline. Client library developers should subclass this to accommodate their particular services.
- authorize_request is a helper method for updating the policy's token and authorizing a request
- on_request allows clients to override the policy's default behavior of authorizing requests. For example, Key Vault clients want to send an unauthorized request to elicit an authentication challenge ([which would look like this](https://github.com/chlowell/azure-sdk-for-python/blob/4d71ae54db6547eeda03d5713b2db818fa3469ae/sdk/keyvault/azure-keyvault-secrets/azure/keyvault/secrets/_shared/challenge_auth_policy.py#L91))
- on_challenge is the method client libraries override to implement challenge handling for their particular service ([ARM, for example](https://github.com/Azure/azure-sdk-for-python/blob/ddaf7529f8fb35c7888e5eeded1e0e0f956709cd/sdk/core/azure-mgmt-core/azure/mgmt/core/policies/_authentication.py#L51))

Reply:
If they want to handle authentication challenges. There's no particular reason to use this policy otherwise, although doing so wouldn't break anything. The policy by itself authorizes requests with bearer tokens just fine.

Reply:
It looks to me we set auth header in both authorize_request & on_request. My understanding is the one in on_request will override authorize_request?

Why user needs to set it twice?

Reply:
on_request is called first when a request arrives. authorize_request is a helper for subclasses to use, so they don't have to handle the credential or cached token ([for example](https://github.com/Azure/azure-sdk-for-python/blob/ddaf7529f8fb35c7888e5eeded1e0e0f956709cd/sdk/core/azure-mgmt-core/azure/mgmt/core/policies/_authentication.py#L51)). Assuming neither method is overridden, authorize_request would overwrite a token applied by on_request. Which is fine, the policy is designed for a scenario like this:
1. on_request authorizes the request
2. policy sends the request to the RP
3. RP responds with a challenge
4. on_challenge authorizes the request with a new token (perhaps by calling authorize_request)
5. policy sends the request to the RP again